File: sdk/python/src/tinyhumans/api/payments.py
from __future__ import annotations

from typing import Any

from ..http import Json
from ._base import ApiNamespace, enc

__all__ = ["PaymentsApi"]


class PaymentsApi(ApiNamespace):
    """Payments: Stripe subscriptions, Coinbase Commerce charges, credit balance
    and top-ups, transaction history, the customer portal, and auto-recharge."""

    def create_coinbase_charge(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Create a Coinbase Commerce charge."""
        return self._http.post("/payments/coinbase/charge", body=body, **kwargs)

    def get_coinbase_charge(
        self,
        gateway_transaction_id: str,
        query: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Json:
        """Get charge status."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get(
            f"/payments/coinbase/charge/{enc(gateway_transaction_id)}",
            query=merged,
            **kwargs,
        )

    def get_auto_recharge(self, **kwargs: Any) -> Json:
        """Get Stripe auto-recharge settings for top-up credits."""
        return self._http.get("/payments/credits/auto-recharge", **kwargs)

    def update_auto_recharge(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Update Stripe auto-recharge settings for top-up credits."""
        return self._http.patch("/payments/credits/auto-recharge", body=body, **kwargs)

    def list_auto_recharge_cards(self, **kwargs: Any) -> Json:
        """List saved Stripe cards for auto recharge."""
        return self._http.get("/payments/credits/auto-recharge/cards", **kwargs)

    def create_auto_recharge_card_setup_intent(self, **kwargs: Any) -> Json:
        """Create a Stripe SetupIntent for adding a saved card."""
        return self._http.post(
            "/payments/credits/auto-recharge/cards/setup-intent", **kwargs
        )

    def update_auto_recharge_card(
        self,
        payment_method_id: str,
        body: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Json:
        """Update a saved Stripe card for auto recharge."""
        return self._http.patch(
            f"/payments/credits/auto-recharge/cards/{enc(payment_method_id)}",
            body=body,
            **kwargs,
        )

    def delete_auto_recharge_card(
        self, payment_method_id: str, **kwargs: Any
    ) -> Json:
        """Delete a saved Stripe card for auto recharge."""
        return self._http.delete(
            f"/payments/credits/auto-recharge/cards/{enc(payment_method_id)}",
            **kwargs,
        )

    def get_credit_balance(self, **kwargs: Any) -> Json:
        """Get the current user's credit balance."""
        return self._http.get("/payments/credits/balance", **kwargs)

    def create_credit_top_up(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Create a credit top-up payment."""
        return self._http.post("/payments/credits/top-up", body=body, **kwargs)

    def get_credit_top_up_cancel(self, **kwargs: Any) -> Json:
        """Handle canceled credit top-up (callback)."""
        return self._http.get("/payments/credits/top-up/cancel", **kwargs)

    def get_credit_top_up_success(
        self, query: dict[str, Any], **kwargs: Any
    ) -> Json:
        """Handle successful credit top-up (callback)."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get(
            "/payments/credits/top-up/success", query=merged, **kwargs
        )

    def list_credit_transactions(
        self, query: dict[str, Any] | None = None, **kwargs: Any
    ) -> Json:
        """Get paginated credit transaction history."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get(
            "/payments/credits/transactions", query=merged, **kwargs
        )

    def get_stripe_checkout_return(
        self, query: dict[str, Any] | None = None, **kwargs: Any
    ) -> Json:
        """Public Stripe Checkout redirect target (browser hand-off)."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get(
            "/payments/stripe/checkout/return", query=merged, **kwargs
        )

    def get_current_plan(self, **kwargs: Any) -> Json:
        """Get current subscription plan for authenticated user."""
        return self._http.get("/payments/stripe/currentPlan", **kwargs)

    def get_stripe_plans(self, **kwargs: Any) -> Json:
        """Get all available subscription plans."""
        return self._http.get("/payments/stripe/plans", **kwargs)

    def create_stripe_portal_session(self, **kwargs: Any) -> Json:
        """Create a Stripe Customer Portal session."""
        return self._http.post("/payments/stripe/portal", **kwargs)

    def get_stripe_portal_return(self, **kwargs: Any) -> Json:
        """Stripe Customer Portal return page."""
        return self._http.get("/payments/stripe/portal/return", **kwargs)

    def purchase_stripe_plan(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Create a Stripe Checkout Session for subscription purchase."""
        return self._http.post("/payments/stripe/purchasePlan", body=body, **kwargs)

File: src/api/payments.rs
//! Payments: Stripe subscriptions, Coinbase Commerce charges, credit balance
//! and top-ups, transaction history, the customer portal, and auto-recharge.

use reqwest::Method;
use serde_json::Value;

use crate::{enc, Error, HttpClient, QueryParam};

/// Typed client for the `/payments/*` routes.
pub struct PaymentsApi<'a> {
    http: &'a HttpClient,
}

impl<'a> PaymentsApi<'a> {
    pub fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    /// Create a Coinbase Commerce charge.
    pub async fn create_coinbase_charge(&self, body: &Value) -> Result<Value, Error> {
        self.http
            .send(
                Method::POST,
                "/payments/coinbase/charge",
                &[],
                Some(body),
                true,
            )
            .await
    }

    /// Get charge status.
    pub async fn get_coinbase_charge(
        &self,
        gateway_transaction_id: &str,
        query: &[QueryParam],
    ) -> Result<Value, Error> {
        let path = format!("/payments/coinbase/charge/{}", enc(gateway_transaction_id));
        self.http.send(Method::GET, &path, query, None, true).await
    }

    /// Get Stripe auto-recharge settings for top-up credits.
    pub async fn get_auto_recharge(&self) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/credits/auto-recharge",
                &[],
                None,
                true,
            )
            .await
    }

    /// Update Stripe auto-recharge settings for top-up credits.
    pub async fn update_auto_recharge(&self, body: &Value) -> Result<Value, Error> {
        self.http
            .send(
                Method::PATCH,
                "/payments/credits/auto-recharge",
                &[],
                Some(body),
                true,
            )
            .await
    }

    /// List saved Stripe cards for auto recharge.
    pub async fn list_auto_recharge_cards(&self) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/credits/auto-recharge/cards",
                &[],
                None,
                true,
            )
            .await
    }

    /// Create a Stripe SetupIntent for adding a saved card.
    pub async fn create_auto_recharge_card_setup_intent(&self) -> Result<Value, Error> {
        self.http
            .send(
                Method::POST,
                "/payments/credits/auto-recharge/cards/setup-intent",
                &[],
                None,
                true,
            )
            .await
    }

    /// Update a saved Stripe card for auto recharge.
    pub async fn update_auto_recharge_card(
        &self,
        payment_method_id: &str,
        body: &Value,
    ) -> Result<Value, Error> {
        let path = format!(
            "/payments/credits/auto-recharge/cards/{}",
            enc(payment_method_id)
        );
        self.http
            .send(Method::PATCH, &path, &[], Some(body), true)
            .await
    }

    /// Delete a saved Stripe card for auto recharge.
    pub async fn delete_auto_recharge_card(&self, payment_method_id: &str) -> Result<Value, Error> {
        let path = format!(
            "/payments/credits/auto-recharge/cards/{}",
            enc(payment_method_id)
        );
        self.http.send(Method::DELETE, &path, &[], None, true).await
    }

    /// Get the current user's credit balance.
    pub async fn get_credit_balance(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/payments/credits/balance", &[], None, true)
            .await
    }

    /// Create a credit top-up payment.
    pub async fn create_credit_top_up(&self, body: &Value) -> Result<Value, Error> {
        self.http
            .send(
                Method::POST,
                "/payments/credits/top-up",
                &[],
                Some(body),
                true,
            )
            .await
    }

    /// Handle canceled credit top-up (callback).
    pub async fn get_credit_top_up_cancel(&self) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/credits/top-up/cancel",
                &[],
                None,
                true,
            )
            .await
    }

    /// Handle successful credit top-up (callback).
    pub async fn get_credit_top_up_success(&self, query: &[QueryParam]) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/credits/top-up/success",
                query,
                None,
                true,
            )
            .await
    }

    /// Get paginated credit transaction history.
    pub async fn list_credit_transactions(&self, query: &[QueryParam]) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/credits/transactions",
                query,
                None,
                true,
            )
            .await
    }

    /// Public Stripe Checkout redirect target (browser hand-off).
    pub async fn get_stripe_checkout_return(&self, query: &[QueryParam]) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/stripe/checkout/return",
                query,
                None,
                true,
            )
            .await
    }

    /// Get current subscription plan for authenticated user.
    pub async fn get_current_plan(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/payments/stripe/currentPlan", &[], None, true)
            .await
    }

    /// Get all available subscription plans.
    pub async fn get_stripe_plans(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/payments/stripe/plans", &[], None, true)
            .await
    }

    /// Create a Stripe Customer Portal session.
    pub async fn create_stripe_portal_session(&self) -> Result<Value, Error> {
        self.http
            .send(Method::POST, "/payments/stripe/portal", &[], None, true)
            .await
    }

    /// Stripe Customer Portal return page.
    pub async fn get_stripe_portal_return(&self) -> Result<Value, Error> {
        self.http
            .send(
                Method::GET,
                "/payments/stripe/portal/return",
                &[],
                None,
                true,
            )
            .await
    }

    /// Create a Stripe Checkout Session for subscription purchase.
    pub async fn purchase_stripe_plan(&self, body: &Value) -> Result<Value, Error> {
        self.http
            .send(
                Method::POST,
                "/payments/stripe/purchasePlan",
                &[],
                Some(body),
                true,
            )
            .await
    }
}
